clang/test/CodeGenSYCL/offload-fp32-div-sqrt.cpp

```diff
@@ -0,0 +1,354 @@
+// RUN: %clang_cc1 %{common_opts_spir64} -ffast-math \
+// RUN: -fno-offload-fp32-prec-div -foffload-fp32-prec-sqrt \
+// RUN: %s -o - | FileCheck --check-prefix ROUNDED-DIV-PREC-SQRT %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -ffast-math \
+// RUN: -fno-offload-fp32-prec-div -fno-offload-fp32-prec-sqrt \
+// RUN: %s -o - | FileCheck --check-prefix ROUNDED-DIV-ROUNDED-SQRT-FAST %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -fno-offload-fp32-prec-div \
+// RUN: -ffp-builtin-accuracy=high %s -o - \
+// RUN: | FileCheck --check-prefix LOW-PREC-DIV %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -ffp-builtin-accuracy=high:div  \
+// RUN: -fno-offload-fp32-prec-div %s -o - \
+// RUN: | FileCheck --check-prefix ROUNDED-DIV %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -fno-offload-fp32-prec-sqrt \
+// RUN: -ffp-builtin-accuracy=high %s -o - \
+// RUN: | FileCheck --check-prefix LOW-PREC-SQRT %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -fno-offload-fp32-prec-div \
+// RUN: -ffp-builtin-accuracy=high:sin %s -o - \
+// RUN: | FileCheck --check-prefix ROUNDED-DIV %s
+
+// RUN: %clang_cc1 %{common_opts_spir64} -fno-offload-fp32-prec-sqrt \
+// RUN: -ffp-builtin-accuracy=high:sin %s -o - \
+// RUN: | FileCheck --check-prefix ROUNDED-SQRT %s
+
+#include "sycl.hpp"
+
+extern "C" SYCL_EXTERNAL float sqrt(float);
+
+using namespace sycl;
+
+int main() {
+  const unsigned array_size = 4;
+  range<1> numOfItems{array_size};
+  float Value1 = .5f;
+  float Value2 = .9f;
+  queue deviceQueue;
+  float *a;
+
+  deviceQueue.submit([&](handler& cgh) {
+    cgh.parallel_for<class KernelSqrt>(numOfItems,
+    [=](id<1> wiID) {
+      // PREC-SQRT: call spir_func float @sqrt(float noundef {{.*}})
+      // ROUNDED-SQRT: call float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // ROUNDED-SQRT-FAST: call reassoc nnan ninf nsz arcp afn float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // PREC-DIV: call spir_func float @sqrt(float noundef {{.*}})
+      // ROUNDED-DIV:  call spir_func float @sqrt(float noundef {{.*}})
+      // ROUNDED-DIV-ROUNDED-SQRT: call float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // PREC-FAST: call reassoc nnan ninf nsz arcp afn spir_func nofpclass(nan inf) float @sqrt(float noundef nofpclass(nan inf) {{.*}})
+      // PREC-SQRT-FAST: call reassoc nnan ninf nsz arcp afn spir_func nofpclass(nan inf) float @sqrt(float noundef nofpclass(nan inf) {{.*}})
+      // ROUNDED-SQRT-PREC-DIV: call reassoc nnan ninf nsz arcp afn float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // ROUNDED-DIV-PREC-SQRT: call reassoc nnan ninf nsz arcp afn spir_func nofpclass(nan inf) float @sqrt(float noundef nofpclass(nan inf) {{.*}})
+      // ROUNDED-DIV-ROUNDED-SQRT-FAST: call reassoc nnan ninf nsz arcp afn float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // LOW-PREC-DIV: call float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT_LOW:[0-9]+]]
```

Comment:
Just want to check if I understand this correctly. In this case we pass: -fno-offload-fp32-prec-div -ffp-builtin-accuracy=high flags. And 1.0 ULP fpbuiltin-error attribute for llvm.fpbuiltin.sqrt.f32 was generated in response of ffp-builtin-accuracy=high flag and we don't expect here precise calculations as high != precise, right?

Reply:
That's correct.